File: NetCoreGameServer/NetCoreGameServer/Data/Model/Character.cs
using System.ComponentModel.DataAnnotations.Schema;
using Newtonsoft.Json;

namespace NetCoreGameServer.Data.Model;

public class Character
{
    public int Id { get; set; }
    public int UserId { get; set; }
    public string Name { get; set; }
    public int Level { get; set; }
    public int ClassId { get; set; }
    public string @Class { get; set; }
    public Gender Gender { get; set; }
    public Core Core { get; set; }
    public int Life { get; set; }
    public int Mana { get; set; }
    public int Experience { get; set; }
    public int StatPoints { get; set; }
    public int AbilityPoints { get; set; }
    public int EquipmentSlots { get; set; }
    public int Kills { get; set; }
    public int Deaths { get; set; }
    public int StatsId { get; set; }

    public Stats Stats { get; set; } = new();

    [JsonIgnore]
    public List<Item> AllItems = new();
    public List<Item> EquippedItems => AllItems.Where(x => x.EquippedItemSlot != null).ToList();
    public List<Item> Equipment => AllItems.Where(x => x.SubType < ItemSubType.Fish && x.EquippedItemSlot == null).ToList();
    public List<Item> Items => AllItems.Where(x => x.SubType >= ItemSubType.Fish && !x.ExpiresAt.HasValue).ToList();
    public List<Item> ActiveGlyphs => AllItems.Where(x => x.ExpiresAt.HasValue).ToList();

    //not persisted

    private const int ItemSlots = 16;
    public Zone? Zone { get; set; }
    [JsonIgnore]
    public long LastRegen { get; set; }

    public bool CanPickItem(Item item)
    {
        return item.IsObject() ? Items.Count < ItemSlots : Equipment.Count < EquipmentSlots;
    }
    
}

File: next/src/models/Character.ts
import { RefObject } from "react";
import Item from "./Item";
import { CharacterStats } from "./Stats";
import { Zone } from "./Zone";
import { EquippedItemSlot } from "./EquippedItem";

export enum CharacterClass {
  Fighter = "Fighter",
  Barbarian = "Barbarian",
  Rogue = "Rogue",
  Magician = "Magician",
  Guardian = "Guardian",
  Samurai = "Samurai",
  Paladin = "Paladin",
  Monk = "Monk",
  Ninja = "Ninja",
  Warlock = "Warlock",
  Headhunter = "Headhunter",
  Alchemist = "Alchemist"
}

export enum Gender {
  Male,
  Female
}

type Character = {
  id: number;
  name: string;
  level: number;
  guild?: string;
  class: CharacterClass;
  gender: Gender;
  equippedItems: Item[];
  equipment: Item[];
  items: Item[];
  life: number;
  mana: number;
  experience: number;
  statPoints: number;
  abilityPoints: number;
  equipmentSlots: number;
  kills: number;
  deaths: number;
  stats: CharacterStats;
  zone: Zone;
  imageRef: RefObject<HTMLDivElement>;
  lastRegenAction: number;
  activeGlyphs: Item[];
}

export default Character;